### DISCUSSION TITLE
How to set default save-as filename for document created by extension

### DISCUSSION THREAD

**Discussion by ptrthomas** [microsoft/vscode-discussions#241] [Extension Development QnA] [answered]
I'm able to create a text-document from scratch and open it like this:

```ts
vscode.workspace.openTextDocument({ language: 'mylang', content: someText })
    .then(document => vscode.window.showTextDocument(document));
```

And then the user can save the file. But the file-name seems to default to the first line of the `content`.

Is there a way to set the file-name (and tab name) of an un-saved file of a TextDocument in the Editor.

**Comment by PEZ** [comment, marked as answer]
The closest I have gotten is to use the signature of `openTextDocument` that takes a `Uri` as an argument.

```typescript
vscode.workspace.openTextDocument(Uri.parse("untitled:Untitled-foo.mylang"))
    .then(document => vscode.window.showTextDocument(document));
```

However there are some differences:

1. You will have to write the text to the document some other way.
2. The file will be ”dirty”, so the user can't just close it without a dialog
3. Possibly more 😄

**Reply by ptrthomas** [reply]
@PEZ thank you so much ! It seems to work well for now.

```ts
                vscode.workspace.openTextDocument(vscode.Uri.parse('untitled:' + utils.workDir() + '/noname.feature'))
                    .then(document => {
                        vscode.window.showTextDocument(document).then(editor => {
                            editor.edit(edit => edit.insert(new vscode.Position(0, 0), someText));
                        });
                    });

```

I would still like an API to do this better - so request the VSC team to consider this.
